### PR TITLE
[MRG] Add warnings for invalid connections

### PR DESCRIPTION
**Pull request by ntolley**
This PR adds checks inside `net.add_connection` to ensure the valid `loc`, `receptor`, `target_type` combinations are specified. Depending on the definition of the cell, certain combinations will break during `simulate_dipole`. For example, pyramidal cells do not have `ampa` synapses so if the user accidentally defines one with `net.add_connection`, then the simulation will break.

While writing this I am realizing there are some issues with how things are organized internally. At the moment we support `proximal`, `distal`, and `soma` as valid location arguments. However, `soma` is handled a bit strangely since it is not defined in the `sect_loc` attribute but still passes as a valid location because it has defined synapses:
https://github.com/jonescompneurolab/hnn-core/blob/4fde7c6f3058cba1fdef96469aa7c86c9d7940f2/hnn_core/cells_default.py#L289

This brings up an important question, should we allow targeting of locations with more precision? For example `loc='apical_1'`. It would be fairly easy to allow, and may make adding to `network_models.py` easier with that sort of flexibility.

### CODE REVIEW

**Comment by jasmainak** [comment]
@ntolley this PR seems pretty easy to bring to MRG. Could be a nice one to tackle to get back into the groove?

**Comment by ntolley** [comment]
> @ntolley this PR seems pretty easy to bring to MRG. Could be a nice one to tackle to get back into the groove?

While it may have taken me some time to "get back into the grove" :smile: , it may be for the better that I came back to this with fresh eyes.

@jasmainak @rythorpe I've implemented a fairly clean system to allow connections to be added through `net.add_connection` that allows `loc` to be the normal targets like `'proximal'` or `'distal'`, but now specific sections like `'apical_tuft'`. It also automatically checks if the synapses exist on that section (the point of this PR).

Since everything goes through `net.add_connection`, it will also be possible to add drives that target specific sections now too (and maybe the tonic bias?). I think this is a really good step towards making the code more generalizable, as now the "grouped sections" like proximal and distal aren't hard-coded and instead derived from the `Cell` object.

**Comment by ntolley** [comment]
Ah I understand now. So that logic is handled here:
https://github.com/jonescompneurolab/hnn-core/blob/b66742d61b447a8bfb01feb20d9911d411985b3f/hnn_core/drives.py#L41

and here:
https://github.com/jonescompneurolab/hnn-core/blob/b66742d61b447a8bfb01feb20d9911d411985b3f/hnn_core/cells_default.py#L339

I think I see why specifying proximal/distal at the cell level could be useful to break away these hard coded dependencies. Perhaps this is getting at @rythorpe's idea of a "constraints" file.

Perhaps the larger issue is simply moving the hard-coded specifications from `drives.py` into a more generalized format.

**Comment by jasmainak** [comment]
Maybe one could iterate over the cell types in `net.cell_types` looking for "proximal" sections ? But if you intend to have connection weights that are cell-type-specific, it's probably better to have the logic outside `add_connection` anyway

**Comment by rythorpe** [comment]
I like this a lot actually @ntolley. I'm definitely in favor of moving away from hardcoded definitions for "proximal" and "distal". These properties are and should be network-specific. Perhaps we could start (probably in a separate PR) by adding a new attribute to `Network` called `Network.connectivity_constraints` of type `dict` that takes the form

```python
Network.connectivity_constraints = {
    'exogenous_drive_targets': {
        'proximal': ['L5_pyr_soma', 'L2_pyr_soma', 'L5_bask_soma', 'L5_bask_soma'],
        'distal': [...]
        },
    'local_net_conn_restrictions': {
        'L5_bask': ['L2_bask', 'L2_pyr'],
        'L2_bask': [...],
        ...
    }
```
By default, `Network.connectivity_constraints` would be left empty (e.g., `{'exogenous_drive_targets': None, 'local_net_conn_restrictions': None}`. The different network models specified in `network_models.py` would then set this attribute when instantiating a specific network configuration.

File: hnn_core/network.py [hunk 1]
```diff
@@ -487,7 +487,10 @@ def add_evoked_drive(self, name, *, mu, sigma, numspikes, location,
         numspikes : int
             Number of spikes at each target cell
         location : str
-            Target location of synapses ('distal' or 'proximal')
+            Target location of synapses. Must be an element of
+            `Cell.sect_loc` such as 'proximal' or 'distal', which defines a
+            group of sections, or an existing section such as 'soma' or
+            'apical_tuft' (defined in `Cell.sections` for all targeted cells)
```
Comment (rythorpe) [inline hnn_core/network.py:493]: +1 for this. Eventually, it'd be nice to allow users to specify which cell sections apply to a given `sect_loc`. I can imagine a few different network changes I'd like to implement for my own project where the location (i.e., 'proximal' or 'distal') isn't necessarily intuitive.

**Review by rythorpe — APPROVED** [review]

File: hnn_core/network_models.py [hunk 2]
```diff
@@ -11,7 +11,8 @@
 from .externals.mne import _validate_type
 
 
-def jones_2009_model(params=None, add_drives_from_params=False):
+def jones_2009_model(params=None, add_drives_from_params=False,
+                     legacy_mode=True):
```
Comment (jasmainak) [inline hnn_core/network_models.py:15]: why did you add this here? We need to set ourselves free of the legacy ...

Comment (ntolley) [reply hnn_core/network_models.py:15]: I promise it's actually towards that goal! `legacy_mode=True` by default in the `Network`. So the old interface for `jones_2009_model()` didn't have the option to turn it on or off.

The reason I added it is because `legacy_mode` needs to be `False` to add a drive that targets a specific section. This is because `legacy_mode=True` creates a connection for _all_ cells, even if they aren't specified in  the weights dict (they just have a weight of zero). Therefore if you try to target `apical_trunk` for example, it'll throw an error because it tries to create a connection to the basket cells when they don't have a section named `apical_trunk`.

Comment (ntolley) [reply hnn_core/network_models.py:15]: Would a good fix be to just make that clearer in the docstring?

Comment (jasmainak) [reply hnn_core/network_models.py:15]: If you plan to deprecate it in the future, the default should be `False` and it should raise a deprecation warning if someone sets it to `True`.

Comment (jasmainak) [reply hnn_core/network_models.py:15]: It sounds like the examples are also using `legacy_mode=True` ? I think the "right" thing to do is to start a separate PR that fixes this problem, then rebase this PR after that is merged.

Comment (rythorpe) [reply hnn_core/network_models.py:15]: If we set default to `False` here, then we'll need to explicitly set `legacy_mode=True` for all of our examples + tests that use this network model in order to maintain consistency. Either that or we just rip the bandaid off and change the outcome of our examples + tests now.

Comment (jasmainak) [reply hnn_core/network_models.py:15]: Btw, I noticed that the new GUI currently loads the "ghost" drives ... probably because of the implicit `legacy_mode=True`

Comment (rythorpe) [reply hnn_core/network_models.py:15]: Ugh. Legacy mode is going to be the death of us.

Comment (ntolley) [reply hnn_core/network_models.py:15]: @jasmainak that's exactly what's happening. Considering this change doesn't require changes to the current code, can we deal with `legacy_mode` in a separate PR?

**Comment by ntolley** [comment]
Just rebased this PR. Everything is good on my end if you guys are happy with the changes!

**Comment by jasmainak** [comment]
Thanks @ntolley ! 🥳 🥳 🥳
